src/scripts/sync_synonym.py

```diff
@@ -0,0 +1,172 @@
+        for pred, alias in sorted(db.alias_relationships(_id)):
+            # - filter all but synonyms: prior in pipeline have already converted other props such as 'label' to syns
+            # todo: is this the best filter? is there more that can appear that should be filtered out? Should instead
+            #  not filter by things we want out, but by the list of synonym preds we want in?
+            if pred == 'rdfs:label':
+                continue
+            tups.add((_id, pred, alias))
+    df = pd.DataFrame(tups, columns=['curie', 'synonym_scope', 'synonym'])
+    return df
+
+
+# TODO
+def sync_synonyms(
+    ontology_db_path: Union[Path, str], mondo_db_path: Union[Path, str], mondo_mappings_path: Union[Path, str],
+    onto_config_path: Union[Path, str], outpath_added: Union[Path, str], outpath_updated: Union[Path, str]
+):
+    """Whenever new synonym or label in upstream source, want to add it to Mondo. If the label is not the Mondo label or
+     existing synonym, it will become a synonym."""
+    # Iterate over mappings
+    mappings_df = pd.read_csv(mondo_mappings_path, sep='\t', comment='#')
+    # todo: refactor filtering by prefixes
+    #  check if this pattern has been used elsewhere; I think it has. and abstract into func for these 3 lines; 'filter prefixes'
+    #  - OR use this?: get_all_owned_terms()
+    #    - it seems long and messy compared to these 3 lines, but is more comprehensive. I could make get_all_owned_terms() a simple 3 line func and then make a separate func for all edge cases?
+    # - filter by source
+    owned_prefix_map: PREFIX_MAP = get_owned_prefix_map(onto_config_path)
+    mappings_df['object_prefix'] = mappings_df['object_id'].apply(lambda x: x.split(':')[0])
+    mappings_df = mappings_df[mappings_df['object_prefix'].isin(owned_prefix_map.keys())]
+    del mappings_df['object_prefix']
+
+    # Query synonyms
+    print(f'Querying synonyms')
+    # - Source
+    # todo: do I need the db's elsewhere? if not, refactor to instantiate them inside the func
+    source_db: SqlImplementation = get_adapter(ontology_db_path)
+    t0 = datetime.now()
+    source_df: pd.DataFrame = _get_synonyms(mappings_df['object_id'].tolist(), source_db)
+    source_df = source_df.rename(columns={'curie': 'source_id'})
+    source_df['mondo_id'] = ''  # todo: do I want to add this at this stage?
+    source_df = source_df[['mondo_id', 'synonym_scope', 'synonym', 'source_id']]  # todo: do I want to add this at this stage?
+    t1 = datetime.now()
+    print(f'- queried source in n seconds: {(t1 - t0).seconds}')
+    # - Mondo
+    mondo_db: SqlImplementation = get_adapter(mondo_db_path)
+    mondo_df: pd.DataFrame = _get_synonyms(mappings_df['subject_id'].tolist(), mondo_db)
+    mondo_df = mondo_df.rename(columns={'curie': 'mondo_id'})
+    t2 = datetime.now()
+    print(f'- queried Mondo (basic info) in n seconds: {(t2 - t1).seconds}')
+    #  - fetch sources for Mondo synonyms
+    # TODO: fetch synonym source
```

Comment:
## Fetching synonym source & performance
This is where I'm stopping for today, and will continue tomorrow.

I don't know if I'll be able to have this first, rough, combined output of multiple types of cases (e.g. -added, -confirmed, -updated, -deleted) by Friday / end of this weekend as hoped.

I was able to quickly and easily get all synonyms and their scope (e.g. exact, broad), querying these for both Mondo and a given source. But this isn't enough.

Basically, without first being able to query the _sources_ on each synonym from Mondo, all of these cases (e.g. -added, -confirmed, -updated, -deleted) will be contain many erroneous rows.

So, since knowing the sources of Mondo synonyms is so important, I've opted to do it now. 

Since we want to try our best to use OAK, and because it's often good, I've taken a first stab at using it.

I first tried using `._axiom_annotations()`. This gives the correct results. However, it is slow, querying on 1 Mondo ID at a time, and will add an estimated 1.5 hours to build time.

I then tried `._axiom_annotations_multi()`, which allows you to query in batch, expecting this to be faster. But I was surprised to see that it is slower. I'm supposing this is due to a lack of indexes. Using it would add about 5 extra hours to build time.

Note also that these are private methods, so I suppose  we shouldn't be using them anyway. But if they had been very convenient and performant, I would've considered asking for a public interface for them.

It seems that I will have to resort to SPARQL, which I don't expect to be very slow anyway.

Reply:
I did this in SPARQL and it was quite fast, just 26 seconds.